doc/content/xapi/migration_overview.md

```diff
@@ -0,0 +1,87 @@
+---
+title: From client RPC migration request to xapi internals
+---
+
+# How the migration request goes through Xen API
+
+## Overview
+
+In this document we will utilize the `VM.pool_migrate` request to illustrate
+the interaction between various components within the XAPI toolstack during
+migration. However this schema can be applied to other requests as well.
+
+Not all parts of the Xapi toolstack are shown here as not all are involved in
+the migration process. For instance you won't see the ***squeezed***
+nor ***mpathalert*** two daemons that belong to the toolstack but don't
+participate in the migration of a VM.
+
+## Anatomy of a VM migration
+
+- Migration is initiated by a Xapi client that sends `VM.pool_migrate`, an RPC XML request.
+- The Xen API server handles this request and dispatches it to the server.
+- The server is generated using `XAPI IDL` and requests are wrapped whitin a
+context, either to be forwared to a host or executed locally. Broadly, the
+context follows RBAC rules. The executed function is related to the message of
+the request (refer to [XenAPI Reference](https://xapi-project.github.io/xen-api/)).
+- In the case of the migration you can refer to *ocaml/idl/datamodel_vm.ml*.
+- The server will dispatch the operation to server helpers, executing the
+operation synchronously or asynchronously and returning the RPC answer.
+- Operations can then be forwarded to another host in the pool or executed
+locally.
+- When executed locally the high-level migration operation is send to the
+*Xenopsd daemon* by posting a message on the message switch.
+*Xenopsd daemon* will split this operation into several *atomic* operations
+that will be run by the *xenopsd backend* in XAPI. The steps involve:
+    - Splitting operations into multiple backend operations.
+    - Posting this operation as an RPC to *xenopsd server*.
+    - The *xenopsd server* then pushes operations onto the message switch.
+    - This allows the *xenopsd backend* to retrieve them and execute them.
+    - interactions between the xenopsd server and its backend occur
+    bidirectionnaly using the broker.
```

Comment:
This is a little different. Xenopsd does indeed have a concept of backends, but a given xenopsd daemon always has just one backend that is compiled together with the frontend.  Call from the frontend (`ocaml/xenopsd/lib/xenops_server.ml`) to the backend are just regular function calls; the message-switch is not involved. The queuing system of operations is part of the frontend, which manages a number of worker threads to execute the operations for different VMs.

Reply:
I think I see what you mean. So at least from the diagram point of view it looks [like this](https://mermaid.ink/svg/pako:eNqFVmFv4kYQ_SsjV1eBBRQIkJxVVeIgUU86JNpGJ9QQcYs94FVsr7u7DtBc_ntnd20MIdX5C971zJs3b3ZmefFCEaEXeJtE7MKYSQ3302W2zICeDx_gjkulYYfAJILS9J1nW1gfgMFiPP8MYcIx06BjpoErUJhFxoBlsJh9af85n4DEfwpU2iE680ZjwXJeLppNaLe_G0cFxp78Kp_v7fZvzs88e_JZKZTPKF-W3rcpVznTYWycaqPTx_drj06a-P63pffqTJU-JFiRV1qKJwx-mozvbm9vW27Z3vFIx8Eg35-oYVO-FyIhJcIn4JlGmbFElaDFeitZHsPSs_nVho2UKTIFsSGlEHL60Fx6NWtHxzE90vk0vht3uxd0qo2IKSqXZIcAhjCsSL4RykgbVUKFLEmgwRG-zjqGwirlxFdj0-hcRm-QsuNCC9hihuZjBIUyFfU_T7_4lzr7_rm-zVMiJxzE6kjjBwxWMSY5SvVATEqAcqcGrqJWtlX0x8volYlh4RRIWca2mJrKGypWree0okJQjk2KSpHdaiPkjklzqg2jMEaq5y5GagdSSRaZragF9v1LnxNmFa_67dLcssQ9hoVGSISBPYClYxgeiRsmX2cw26a6Rkv4ExKHy3wupTkDszEnJoF3-m2PmchtLe4lyxQxTcHtnQqN0CgD07dcuHo0_y_tE2D41QSfC5oxG5HQCDIa0EmDlKYSHVYD7SxXttk2LLQVcpqkartSO06H6qzuJ21oXSOYMkxFdtpxVv6j-8PDzJUC3DqAPwqkCfT4aAnOCxX_Mhc5iAyE3HaIUcck0XHUok6YMKV4aEmVe2UDGuUWVpMI_I0UlEQWvemjLWr4GVSecA0x38aQ4DMlL-jYak4hKXMBTIuUh8poCufu56vz8CCIfsKzJ-pjmtChSHNOk0bzFImsOFqvaUoRsRrquLAv9QAsBS3tzchnYUjamboRGA3-yPwqLSQ6rzchSkEsil_ulXp8KkEXE2gkfE2OoZaJPUbmjLpA1HWVzKv4QBI9cyWMzAQK9QYJeqRRnUI3Zc8d62l7e3v3dtr2z4b_hC41c01Vt5Ppf9P7sTm9biK4-VLQeo1VG5c6vt_r5co5lplhgiG5rQwu5fX7j-DP0zv1rpPrjt-51ryWl6JMGY_o_n8xIEuPEkpJsYBeI9ywItGma17JlNGt8NchC71AywJbXpFHNF6mnFGzpV6woWuQdnOW_S1EWhnR0gtevL0XDEajTm_Q6970r0dXo-vex5Z38IKrQedq2B30h9f9br83uBkMXlvevxag2_nYG3ZvRn3yGfRGV_2blocRp4rO3D8W-8fl9T-6EN9m)
Not sure if you can see the link...

Reply:
Note that the `message-switch` is a separate daemon; it's not part of xenopsd. It is used for communication between the various toolstack daemons on a host.

Reply:
Oh yes right. I will show that in next graph. Thx for the enlightenment.